[inductor] Modify the rules for freezing the layout of x.unwrap_view() in convert_to_reinterpret_view

jiayisunx opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122760



Fix https://github.com/pytorch/pytorch/issues/121607

Modify the rules for freezing the layout of `x.unwrap_view()` in `convert_to_reinterpret_view`: If any read of `x.unwrap_view()` is in channels_last format, freeze the layout of `x.unwrap_view()` to channels_last format. 


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

jiayisunx committed:
[inductor] Modify the rules for freezing layout of x.unwrap_view() in convert_to_reinterpret_view

[ghstack-poisoned]

jiayisunx committed:
Update on "[inductor] Modify the rules for freezing the layout of x.unwrap_view() in convert_to_reinterpret_view"




Fix https://github.com/pytorch/pytorch/issues/121607

Modify the rules for freezing the layout of `x.unwrap_view()` in `convert_to_reinterpret_view`: If any read of `x.unwrap_view()` is in channels_last format, freeze the layout of `x.unwrap_view()` to channels_last format. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jiayisunx committed:
Update on "[inductor] Modify the rules for freezing the layout of x.unwrap_view() in convert_to_reinterpret_view"




Fix https://github.com/pytorch/pytorch/issues/121607

Modify the rules for freezing the layout of `x.unwrap_view()` in `convert_to_reinterpret_view`: If any read of `x.unwrap_view()` is in channels_last format, freeze the layout of `x.unwrap_view()` to channels_last format. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jiayisunx committed:
Update on "[inductor] Modify the rules for freezing the layout of x.unwrap_view() in convert_to_reinterpret_view"




Fix https://github.com/pytorch/pytorch/issues/121607

Modify the rules for freezing the layout of `x.unwrap_view()` in `convert_to_reinterpret_view`: If any read of `x.unwrap_view()` is in channels_last format, freeze the layout of `x.unwrap_view()` to channels_last format. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jiayisunx committed:
Update on "[inductor] Modify the rules for freezing the layout of x.unwrap_view() in convert_to_reinterpret_view"




Fix https://github.com/pytorch/pytorch/issues/121607

Modify the rules for freezing the layout of `x.unwrap_view()` in `convert_to_reinterpret_view`: If any read of `x.unwrap_view()` is in channels_last format, freeze the layout of `x.unwrap_view()` to channels_last format. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jiayisunx committed:
Update on "[inductor] Modify the rules for freezing the layout of x.unwrap_view() in convert_to_reinterpret_view"




Fix https://github.com/pytorch/pytorch/issues/121607

Modify the rules for freezing the layout of `x.unwrap_view()` in `convert_to_reinterpret_view`: If any read of `x.unwrap_view()` is in channels_last format, freeze the layout of `x.unwrap_view()` to channels_last format. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jiayisunx committed:
Update on "[inductor] Modify the rules for freezing the layout of x.unwrap_view() in convert_to_reinterpret_view"




Fix https://github.com/pytorch/pytorch/issues/121607

Modify the rules for freezing the layout of `x.unwrap_view()` in `convert_to_reinterpret_view`: If any read of `x.unwrap_view()` is in channels_last format, freeze the layout of `x.unwrap_view()` to channels_last format. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel commented on torch/_inductor/ir.py:
```diff
@@ -3978,7 +3978,28 @@ def convert_to_reinterpret_view(cls, x):
+                isinstance(buf.layout, FixedLayout)
+                and buf.layout.is_channels_last_contiguous()
+                for buf in reads_bufs
+            )
```
nit: add some comments for the heuristic to decide the layout of `x.unwrap_view()`?

jiayisunx replied:
I have added comments for this heuristic. Thanks.

jiayisunx committed:
Update on "[inductor] Modify the rules for freezing the layout of x.unwrap_view() in convert_to_reinterpret_view"




Fix https://github.com/pytorch/pytorch/issues/121607

Modify the rules for freezing the layout of `x.unwrap_view()` in `convert_to_reinterpret_view`: If any read of `x.unwrap_view()` is in channels_last format, freeze the layout of `x.unwrap_view()` to channels_last format. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jiayisunx committed:
Update on "[inductor] Modify the rules for freezing the layout of x.unwrap_view() in convert_to_reinterpret_view"




Fix https://github.com/pytorch/pytorch/issues/121607

Modify the rules for freezing the layout of `x.unwrap_view()` in `convert_to_reinterpret_view`: If any read of `x.unwrap_view()` is in channels_last format, freeze the layout of `x.unwrap_view()` to channels_last format. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jiayisunx commented on torch/_inductor/ir.py:
```diff
@@ -3978,7 +3978,28 @@ def convert_to_reinterpret_view(cls, x):
+                for buf in reads_bufs
+            )
+        ):
+            x.unwrap_view().freeze_layout_with_same_order(
```
Done, thanks.

jiayisunx committed:
Update on "[inductor] Modify the rules for freezing the layout of x.unwrap_view() in convert_to_reinterpret_view"




Fix https://github.com/pytorch/pytorch/issues/121607

Modify the rules for freezing the layout of `x.unwrap_view()` in `convert_to_reinterpret_view`: If any read of `x.unwrap_view()` is in channels_last format, freeze the layout of `x.unwrap_view()` to channels_last format. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jiayisunx committed:
Update on "[inductor] Modify the rules for freezing the layout of x.unwrap_view() in convert_to_reinterpret_view"




Fix https://github.com/pytorch/pytorch/issues/121607

Modify the rules for freezing the layout of `x.unwrap_view()` in `convert_to_reinterpret_view`: If any read of `x.unwrap_view()` is in channels_last format, freeze the layout of `x.unwrap_view()` to channels_last format. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jiayisunx committed:
Update on "[inductor] Modify the rules for freezing the layout of x.unwrap_view() in convert_to_reinterpret_view"




Fix https://github.com/pytorch/pytorch/issues/121607

Modify the rules for freezing the layout of `x.unwrap_view()` in `convert_to_reinterpret_view`: If any read of `x.unwrap_view()` is in channels_last format, freeze the layout of `x.unwrap_view()` to channels_last format. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jiayisunx committed:
Update on "[inductor] Modify the rules for freezing the layout of x.unwrap_view() in convert_to_reinterpret_view"




Fix https://github.com/pytorch/pytorch/issues/121607

Modify the rules for freezing the layout of `x.unwrap_view()` in `convert_to_reinterpret_view`: If any read of `x.unwrap_view()` is in channels_last format, freeze the layout of `x.unwrap_view()` to channels_last format. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jiayisunx committed:
Update on "[inductor] Modify the rules for freezing the layout of x.unwrap_view() in convert_to_reinterpret_view"




Fix https://github.com/pytorch/pytorch/issues/121607

Modify the rules for freezing the layout of `x.unwrap_view()` in `convert_to_reinterpret_view`: If any read of `x.unwrap_view()` is in channels_last format, freeze the layout of `x.unwrap_view()` to channels_last format. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jiayisunx committed:
Update on "[inductor] Modify the rules for freezing the layout of x.unwrap_view() in convert_to_reinterpret_view"




Fix https://github.com/pytorch/pytorch/issues/121607

Modify the rules for freezing the layout of `x.unwrap_view()` in `convert_to_reinterpret_view`: If any read of `x.unwrap_view()` is in channels_last format, freeze the layout of `x.unwrap_view()` to channels_last format. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jgong5 commented on torch/_inductor/ir.py:
```diff
@@ -4011,7 +4011,31 @@ def convert_to_reinterpret_view(cls, x):
         # make_loader() inlines the computation
-        x.unwrap_view().freeze_layout()
+        x_unwrap_view = x.unwrap_view()
+        x_unwrap_view_node = V.graph.get_buffer(
```
naming it `x_unwrap_view_fx_node` is more clear.

jiayisunx replied:
Done, thanks!

jgong5 commented on torch/_inductor/ir.py:
```diff
@@ -4011,7 +4011,31 @@ def convert_to_reinterpret_view(cls, x):
+        x_unwrap_view_node = V.graph.get_buffer(
+            x_unwrap_view.get_name()
+        ).get_origin_node()
+        # If the node of x.unwrap_view() has meta tensor and the meta tensor is in CL format,
+        # freeze the layout of x.unwrap_view() as CL format.
```
```suggestion
        # Prefer channels last format according to how the format is set from eager.
```

jiayisunx replied:
Done, thanks!

jgong5 reviewed (approved):
LGTM. Minor suggestion on naming/wording.

leslie-fang-intel commented on torch/_inductor/ir.py:
```diff
@@ -4011,7 +4011,31 @@ def convert_to_reinterpret_view(cls, x):
+        x_unwrap_view = x.unwrap_view()
+        x_unwrap_view_node = V.graph.get_buffer(
+            x_unwrap_view.get_name()
+        ).get_origin_node()
```
`get_origin_node` may not always work as comment in https://github.com/pytorch/pytorch/blob/1fd9e320ea415c4b221799863a334c842bdb6ff2/torch/_inductor/graph.py#L1254-L1261 But here since we check and decide the CL format only when `get_origin_node` exists. I think it should be ok.

jiayisunx committed:
Update on "[inductor] Modify the rules for freezing the layout of x.unwrap_view() in convert_to_reinterpret_view"




Fix https://github.com/pytorch/pytorch/issues/121607

Modify the rules for freezing the layout of `x.unwrap_view()` in `convert_to_reinterpret_view`: If any read of `x.unwrap_view()` is in channels_last format, freeze the layout of `x.unwrap_view()` to channels_last format. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jiayisunx commented:
@jansel , could you please review this PR? Thanks.

jiayisunx commented:
@pytorchbot merge

jiayisunx commented:
@pytorchbot merge
